File: packages/unswallow/src/recover.ts
import { randomUUID } from 'node:crypto';
import type { RawProviderResponse, ToolCallEntry, ToolEnvelope } from './types';

const TOOL_OPEN_AT = /<(tool_call|tool_calls)\b[^>]*>/iy;
const FUNC_OPEN_AT = /<function=([^>]*)>/iy;
const PARAMETER = /<parameter=([^>]*)>([\s\S]*?)<\/parameter\s*>/gi;
const MAX_ENVELOPE_LENGTH = 20000;
export const MAX_ENVELOPES = 32;

export interface LocatedEnvelope extends ToolEnvelope {
  start: number;
  end: number;
}

export interface ExtractionResult {
  envelopes: LocatedEnvelope[];
  capped: boolean;
}

function findClose(text: string, tag: string, from: number): { start: number; end: number } | null {
  const close = new RegExp(`</${tag}\\s*>`, 'i');
  close.lastIndex = from;
  const m = close.exec(text);
  if (!m) return null;
  return { start: m.index, end: m.index + m[0].length };
}

function coerceScalar(v: string): unknown {
  const t = v.trim();
  if (/^-?\d+$/.test(t)) return parseInt(t, 10);
  if (/^-?\d+\.\d+$/.test(t)) return parseFloat(t);
  if (t === 'true') return true;
  if (t === 'false') return false;
  if (t === 'null') return null;
  return t;
}

export function validateEnvelopeShape(obj: unknown): {
  name: string;
  arguments: Record<string, unknown>;
  argumentsFromString?: boolean;
} | null {
  if (typeof obj !== 'object' || obj === null || Array.isArray(obj)) return null;
  const o = obj as Record<string, unknown>;
  const fn =
    typeof o.function === 'object' && o.function !== null && !Array.isArray(o.function)
      ? (o.function as Record<string, unknown>)
      : o;
  if (typeof fn.name !== 'string' || fn.name.trim().length === 0 || fn.name.length > 200) {
    return null;
  }
  let args = fn.arguments;
  let fromString = false;
  if (typeof args === 'string') {
    try {
      args = JSON.parse(args);
      fromString = true;
    } catch {
      return null;
    }
  }
  if (typeof args !== 'object' || args === null || Array.isArray(args)) return null;
  return {
    name: fn.name.trim(),
    arguments: args as Record<string, unknown>,
    argumentsFromString: fromString,
  };
}

function tryQwenAt(text: string, tag: string, openStart: number, openLen: number): LocatedEnvelope | null {
  const innerStart = openStart + openLen;
  const close = findClose(text, tag, innerStart);
  if (!close) return null;
  const inner = text.slice(innerStart, close.start).trim();
  const raw = text.slice(openStart, close.end);
  if (inner.length === 0 || raw.length > MAX_ENVELOPE_LENGTH) return null;
  const attrName = /name\s*=\s*"([^"]*)"/i.exec(text.slice(openStart, innerStart))?.[1]?.trim();
  if (inner.startsWith('{') || inner.startsWith('[')) {
    let obj: unknown;
    try {
      obj = JSON.parse(inner);
    } catch {
      return null;
    }
    const shape = validateEnvelopeShape(obj);
    if (!shape) return null;
    return {
      name: shape.name,
      arguments: shape.arguments,
      argumentsFromString: shape.argumentsFromString,
      raw,
      format: 'qwen-xml',
      start: openStart,
      end: close.end,
    };
  }
  if (attrName) {
    const params: Record<string, unknown> = {};
    PARAMETER.lastIndex = 0;
    let pm: RegExpExecArray | null;
    let any = false;
    while ((pm = PARAMETER.exec(inner)) !== null) {
      params[pm[1].trim()] = coerceScalar(pm[2]);
      any = true;
    }
    if (any) {
      return { name: attrName, arguments: params, raw, format: 'qwen-xml', start: openStart, end: close.end };
    }
  }
  return null;
}

function tryFunctionAt(text: string, name: string, openStart: number, openLen: number): LocatedEnvelope | null {
  const clean = name.trim().replace(/^["']|["']$/g, '');
  if (!clean) return null;
  const innerStart = openStart + openLen;
  const close = findClose(text, 'function', innerStart);
  if (!close) return null;
  const inner = text.slice(innerStart, close.start).trim();
  const raw = text.slice(openStart, close.end);
  if (raw.length > MAX_ENVELOPE_LENGTH) return null;
  const params: Record<string, unknown> = {};
  PARAMETER.lastIndex = 0;
  let pm: RegExpExecArray | null;
  let any = false;
  while ((pm = PARAMETER.exec(inner)) !== null) {
    params[pm[1].trim()] = coerceScalar(pm[2]);
    any = true;
  }
  if (any) {
    return { name: clean, arguments: params, raw, format: 'function-xml', start: openStart, end: close.end };
  }
  if (inner.startsWith('{')) {
    try {
      const shape = validateEnvelopeShape(JSON.parse(inner));
      if (shape) {
        return {
          name: shape.name,
          arguments: shape.arguments,
          argumentsFromString: shape.argumentsFromString,
          raw,
          format: 'function-xml',
          start: openStart,
          end: close.end,
        };
      }
    } catch {
      return null;
    }
  }
  return null;
}

function tryJsonAt(text: string, braceIndex: number): LocatedEnvelope | null {
  const end = scanBalancedJson(text, braceIndex);
  if (end === -1) return null;
  const raw = text.slice(braceIndex, end);
  if (raw.length > MAX_ENVELOPE_LENGTH) return null;
  let shape: ReturnType<typeof validateEnvelopeShape> = null;
  try {
    shape = validateEnvelopeShape(JSON.parse(raw));
  } catch {
    return null;
  }
  if (!shape) return null;
  return {
    name: shape.name,
    arguments: shape.arguments,
    argumentsFromString: shape.argumentsFromString,
    raw,
    format: 'json',
    start: braceIndex,
    end,
  };
}

export function extractAllEnvelopes(text: string): ExtractionResult {
  const envelopes: LocatedEnvelope[] = [];
  let capped = false;
  const n = text.length;
  let pos = 0;
  let guard = 0;
  while (pos < n) {
    if (guard++ > n + 16) break;
    if (envelopes.length >= MAX_ENVELOPES) {
      capped = true;
      break;
    }
    const lt = text.indexOf('<', pos);
    const brace = text.indexOf('{', pos);
    let next = -1;
    if (lt === -1) next = brace;
    else if (brace === -1) next = lt;
    else next = Math.min(lt, brace);
    if (next === -1) break;
    if (text[next] === '{') {
      const hit = tryJsonAt(text, next);
      if (hit) {
        envelopes.push(hit);
        pos = hit.end;
      } else {
        pos = next + 1;
      }
      continue;
    }
    TOOL_OPEN_AT.lastIndex = next;
    const tool = TOOL_OPEN_AT.exec(text);
    if (tool) {
      const hit = tryQwenAt(text, tool[1], next, tool[0].length);
      if (hit) {
        envelopes.push(hit);
        pos = hit.end;
      } else {
        pos = next + tool[0].length;
      }
      continue;
    }
    FUNC_OPEN_AT.lastIndex = next;
    const func = FUNC_OPEN_AT.exec(text);
    if (func) {
      const hit = tryFunctionAt(text, func[1] ?? '', next, func[0].length);
      if (hit) {
        envelopes.push(hit);
        pos = hit.end;
      } else {
        pos = next + func[0].length;
      }
      continue;
    }
    pos = next + 1;
  }
  return { envelopes, capped };
}

function scanBalancedJson(text: string, start: number): number {
  let depth = 0;
  let inString = false;
  let escaped = false;
  for (let i = start; i < text.length; i++) {
    const ch = text[i];
    if (inString) {
      if (escaped) escaped = false;
      else if (ch === '\\') escaped = true;
      else if (ch === '"') inString = false;
      continue;
    }
    if (ch === '"') inString = true;
    else if (ch === '{' || ch === '[') depth++;
    else if (ch === '}' || ch === ']') {
      depth--;
      if (depth === 0) return i + 1;
    }
  }
  return -1;
}

export function extractEnvelope(text: string): ToolEnvelope | null {
  const found = extractAllEnvelopes(text).envelopes[0];
  if (!found) return null;
  const { start: _s, end: _e, ...envelope } = found;
  return envelope;
}

export function buildToolCallsEntry(
  name: string,
  argumentsObj: Record<string, unknown>
): ToolCallEntry {
  return {
    id: `call_${randomUUID().replace(/-/g, '').slice(0, 24)}`,
    type: 'function',
    function: { name, arguments: JSON.stringify(argumentsObj) },
  };
}

export function applyRecoveryMany(
  response: RawProviderResponse,
  calls: Array<{ name: string; arguments: Record<string, unknown> }>
): RawProviderResponse {
  const clone = structuredClone(response);
  const choice = clone.choices[0];
  choice.message.tool_calls = calls.map((c) => buildToolCallsEntry(c.name, c.arguments));
  if (choice.finish_reason === 'stop') {
    choice.finish_reason = 'tool_calls';
  }
  return clone;
}

export function applyRecoveryToResponse(
  response: RawProviderResponse,
  name: string,
  argumentsObj: Record<string, unknown>
): RawProviderResponse {
  return applyRecoveryMany(response, [{ name, arguments: argumentsObj }]);
}

File: packages/python/unswallow/recover.py
from __future__ import annotations

import copy
import json
import re
import uuid
from typing import Any, Dict, Optional

from .types import ToolEnvelope

TOOL_OPEN_AT = re.compile(r"<(tool_call|tool_calls)\b[^>]*>", re.I)
FUNC_OPEN_AT = re.compile(r"<function=([^>]*)>", re.I)
PARAMETER = re.compile(r"<parameter=([^>]*)>([\s\S]*?)</parameter\s*>", re.I)
NAME_ATTR = re.compile(r'name\s*=\s*"([^"]*)"', re.I)
MAX_ENVELOPE_LENGTH = 20000
MAX_ENVELOPES = 32


def _find_close(text: str, tag: str, start: int) -> Optional[Tuple[int, int]]:
    m = re.search(r"</" + re.escape(tag) + r"\s*>", text[start:], re.I)
    if not m:
        return None
    return start + m.start(), start + m.end()


def _coerce_scalar(v: str) -> Any:
    t = v.strip()
    if re.match(r"^-?\d+$", t):
        return int(t)
    if re.match(r"^-?\d+\.\d+$", t):
        return float(t)
    if t == "true":
        return True
    if t == "false":
        return False
    if t == "null":
        return None
    return t


def validate_envelope_shape(obj: Any) -> Optional[Dict[str, Any]]:
    if not isinstance(obj, dict):
        return None
    fn = obj.get("function")
    if not isinstance(fn, dict):
        fn = obj
    name = fn.get("name")
    if not isinstance(name, str) or not name.strip() or len(name) > 200:
        return None
    args = fn.get("arguments")
    arguments_from_string = False
    if isinstance(args, str):
        try:
            args = json.loads(args)
            arguments_from_string = True
        except ValueError:
            return None
    if not isinstance(args, dict):
        return None
    return {"name": name.strip(), "arguments": args, "arguments_from_string": arguments_from_string}


def _try_qwen_at(text: str, tag: str, open_start: int, open_len: int) -> Optional[ToolEnvelope]:
    inner_start = open_start + open_len
    close = _find_close(text, tag, inner_start)
    if not close:
        return None
    inner = text[inner_start : close[0]].strip()
    raw = text[open_start : close[1]]
    if not inner or len(raw) > MAX_ENVELOPE_LENGTH:
        return None
    attr_m = NAME_ATTR.search(text[open_start:inner_start])
    attr_name = attr_m.group(1).strip() if attr_m else None
    if inner.startswith("{") or inner.startswith("["):
        try:
            shape = validate_envelope_shape(json.loads(inner))
        except ValueError:
            shape = None
        if shape:
            return ToolEnvelope(
                name=shape["name"],
                arguments=shape["arguments"],
                arguments_from_string=shape["arguments_from_string"],
                raw=raw,
                format="qwen-xml",
                start=open_start,
                end=close[1],
            )
    if attr_name:
        params = {}
        any_param = False
        for pm in PARAMETER.finditer(inner):
            params[pm.group(1).strip()] = _coerce_scalar(pm.group(2))
            any_param = True
        if any_param:
            return ToolEnvelope(name=attr_name, arguments=params, raw=raw, format="qwen-xml", start=open_start, end=close[1])
    return None


def _try_function_at(text: str, name: str, open_start: int, open_len: int) -> Optional[ToolEnvelope]:
    clean = name.strip().strip('"').strip("'")
    if not clean:
        return None
    inner_start = open_start + open_len
    close = _find_close(text, "function", inner_start)
    if not close:
        return None
    inner = text[inner_start : close[0]].strip()
    raw = text[open_start : close[1]]
    if len(raw) > MAX_ENVELOPE_LENGTH:
        return None
    params = {}
    any_param = False
    for pm in PARAMETER.finditer(inner):
        params[pm.group(1).strip()] = _coerce_scalar(pm.group(2))
        any_param = True
    if any_param:
        return ToolEnvelope(name=clean, arguments=params, raw=raw, format="function-xml", start=open_start, end=close[1])
    if inner.startswith("{"):
        try:
            shape = validate_envelope_shape(json.loads(inner))
        except ValueError:
            shape = None
        if shape:
            return ToolEnvelope(
                name=shape["name"],
                arguments=shape["arguments"],
                arguments_from_string=shape["arguments_from_string"],
                raw=raw,
                format="function-xml",
                start=open_start,
                end=close[1],
            )
    return None


def _try_json_at(text: str, brace_index: int) -> Optional[ToolEnvelope]:
    end = _scan_balanced_json(text, brace_index)
    if end == -1:
        return None
    raw = text[brace_index:end]
    if len(raw) > MAX_ENVELOPE_LENGTH:
        return None
    try:
        shape = validate_envelope_shape(json.loads(raw))
    except ValueError:
        return None
    if not shape:
        return None
    return ToolEnvelope(
        name=shape["name"],
        arguments=shape["arguments"],
        arguments_from_string=shape["arguments_from_string"],
        raw=raw,
        format="json",
        start=brace_index,
        end=end,
    )


def extract_all_envelopes(text: str):
    envelopes = []
    capped = False
    n = len(text)
    pos = 0
    guard = 0
    while pos < n:
        guard += 1
        if guard > n + 16:
            break
        if len(envelopes) >= MAX_ENVELOPES:
            capped = True
            break
        lt = text.find("<", pos)
        brace = text.find("{", pos)
        if lt == -1:
            nxt = brace
        elif brace == -1:
            nxt = lt
        else:
            nxt = lt if lt < brace else brace
        if nxt == -1:
            break
        if text[nxt] == "{":
            hit = _try_json_at(text, nxt)
            if hit:
                envelopes.append(hit)
                pos = hit.end
            else:
                pos = nxt + 1
            continue
        m = TOOL_OPEN_AT.match(text, nxt)
        if m:
            hit = _try_qwen_at(text, m.group(1), nxt, len(m.group(0)))
            if hit:
                envelopes.append(hit)
                pos = hit.end
            else:
                pos = nxt + len(m.group(0))
            continue
        m = FUNC_OPEN_AT.match(text, nxt)
        if m:
            hit = _try_function_at(text, m.group(1) or "", nxt, len(m.group(0)))
            if hit:
                envelopes.append(hit)
                pos = hit.end
            else:
                pos = nxt + len(m.group(0))
            continue
        pos = nxt + 1
    return envelopes, capped


def _scan_balanced_json(text: str, start: int) -> int:
    depth = 0
    in_string = False
    escaped = False
    i = start
    while i < len(text):
        ch = text[i]
        if in_string:
            if escaped:
                escaped = False
            elif ch == "\\":
                escaped = True
            elif ch == '"':
                in_string = False
        else:
            if ch == '"':
                in_string = True
            elif ch in "{[":
                depth += 1
            elif ch in "}]":
                depth -= 1
                if depth == 0:
                    return i + 1
        i += 1
    return -1


def extract_envelope(text: str) -> Optional[ToolEnvelope]:
    found = extract_all_envelopes(text)[0]
    if not found:
        return None
    first = found[0]
    return ToolEnvelope(
        name=first.name,
        arguments=first.arguments,
        raw=first.raw,
        format=first.format,
        arguments_from_string=first.arguments_from_string,
    )


def build_tool_calls_entry(name: str, arguments: Dict[str, Any]) -> Dict[str, Any]:
    return {
        "id": "call_" + uuid.uuid4().hex[:24],
        "type": "function",
        "function": {"name": name, "arguments": json.dumps(arguments)},
    }


def apply_recovery_many(response: Dict[str, Any], calls) -> Dict[str, Any]:
    clone = copy.deepcopy(response)
    choice = clone["choices"][0]
    choice["message"]["tool_calls"] = [build_tool_calls_entry(c["name"], c["arguments"]) for c in calls]
    if choice.get("finish_reason") == "stop":
        choice["finish_reason"] = "tool_calls"
    return clone


def apply_recovery_to_response(response: Dict[str, Any], name: str, arguments: Dict[str, Any]) -> Dict[str, Any]:
    return apply_recovery_many(response, [{"name": name, "arguments": arguments}])